x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/SpikeAndDipDetector.java

```diff
@@ -0,0 +1,174 @@
+
+        boolean includes(int i) {
+            return i >= startExcluded && i < endExcluded;
+        }
+    }
+
+    private int argmax(double[] values, int start, int end, boolean negate) {
+        int argmax = 0;
+        double max = negate ? -values[0] : values[0];
+        for (int i = 1; i < values.length; i++) {
+            double value = negate ? -values[i] : values[i];
+            if (value > max) {
+                argmax = i;
+                max = value;
+            }
+        }
+        return argmax;
+    }
+
+    private double average(double[] values, int start, int end, boolean negate) {
+        double sum = 0.0;
+        for (int i = start; i < end; i++) {
+            sum += negate ? -values[i] : values[i];
+        }
+        return sum / (double) (end - start);
+    }
+
+    private SpikeOrDip findSpikeOrDip(double[] values, int extent, boolean negate) {
+
+        extent = Math.min(extent, values.length - 1);
+
+        final int argmax = argmax(values, 0, values.length, negate);
+
+        // Find the maximum average interval of width extent which includes argmax.
+        int maxStart = Math.max(0, argmax + 1 - extent);
+        int maxEnd = Math.min(maxStart + extent, values.length);
+        double maxAvg = average(values, maxStart, maxEnd, negate);
+        for (int start = maxStart + 1; start <= argmax; start++) {
+            if (start + extent >= values.length) {
+                break;
+            }
+            double average = average(values, start, start + extent, negate);
+            if (average > maxAvg) {
+                maxStart = start;
+                maxAvg = average;
+            }
+        }
+
+        return new SpikeOrDip(argmax, maxStart, maxStart + extent);
+    }
+
+    private interface ExcludedPredicate {
+        boolean exclude(int index);
+    }
+
+    private double[] removeIf(ExcludedPredicate should, double[] values) {
+        int numKept = 0;
+        for (int i = 0; i < values.length; i++) {
+            if (should.exclude(i) == false) {
+                numKept++;
+            }
+        }
+        double[] newValues = new double[numKept];
+        int j = 0;
+        for (int i = 0; i < values.length; i++) {
+            if (should.exclude(i) == false) {
+                newValues[j++] = values[i];
+            }
+        }
+        return newValues;
+    }
+
+    private final int numValues;
+    private final int dipIndex;
+    private final int spikeIndex;
+    private final double dipValue;
+    private final double spikeValue;
+    private final KDE spikeTestKDE;
+    private final KDE dipTestKDE;
+
+    SpikeAndDipDetector(double[] values) {
+
+        numValues = values.length;
+
+        if (values.length < 4) {
+            dipIndex = -1;
+            spikeIndex = -1;
+            dipValue = Double.NaN;
+            spikeValue = Double.NaN;
+            spikeTestKDE = new KDE(values, 0.8);
+            dipTestKDE = new KDE(values, 0.8);
+            return;
```

Comment:
seems like `spikeTestKDE` & `dipTestKDE` should be null here. We should never call it when the value length is < 4 correct?

Reply:
I thought it was better to have them initialised to minimise any risk of introducing NPE in the future, but they can definitely be null in current usage. I changed over as you suggest.